### ISSUE TITLE
Permitted assignment to read-only capture in not-mutable lambda

### ISSUE THREAD

**Issue by Fedr** [llvm/llvm-project#95081]
This program
```
struct B {
    int i;
};

int foo() {
    auto [x] = B{1};
    [x]() {
        x = 2;
    }();
    return x;
}
```
is invalid because not-mutable lambda modifies its read-only capture `x`, and it is properly rejected by GCC and MSVC, but Clang erroneously admits it. Online demo: https://gcc.godbolt.org/z/46o6jad8n

**Comment by shafik** [comment]
If we use `x=x` instead, we obtain the diagnostic we expect: https://gcc.godbolt.org/z/MbjY4a1j8

Which is `err_lambda_decl_ref_not_modifiable_lvalue` emitted from here:

https://github.com/llvm/llvm-project/blob/a13bc9714a6bfb766693aa7900217f6f9be6f25d/clang/lib/Sema/SemaExpr.cpp#L13507

It looks like when running `E->isModifiableLvalue(...)` we end up eventually here:

https://github.com/llvm/llvm-project/blob/a13bc9714a6bfb766693aa7900217f6f9be6f25d/clang/lib/AST/ExprClassification.cpp#L677-L680

In the OP's case `CT.isConstQualified()` is `false` and I don't see any other point where it could/should catch this case.

**Comment by zygoloid** [comment]
We're computing the wrong type for the `DeclRefExpr`. Eg, we reject-valid on this:
```c++
void f(int&) = delete;
void f(const int&);

int arr[1];
void foo() {
    auto [x] = arr;
    [x]() { f(x); }();
}
```

**Comment by shafik** [comment]
At this point:

https://github.com/llvm/llvm-project/blob/7cff05ada05e87408966d56b4c1675033187ff5c/clang/lib/Sema/SemaExpr.cpp#L3332

for the `x=x` the result of `D->getKind()` is `Var` but is `Binding` for the OP's case.

So it looks like we skip the `getCapturedDeclRefType(cast<VarDecl>(VD), Loc);` in the OP's case. 

When we do the lookup here: 

https://github.com/llvm/llvm-project/blob/8b18684708e4fbdd521b8df9a16b564315ff2c43/clang/lib/Sema/SemaDecl.cpp#L1281-L1283

We are missing the capture in the OP's case.
